src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    internal class SocketSenderPool : IDisposable
+    {
+        private const int MaxQueueSize = 1024; // REVIEW: Is this good enough?
+
+        private readonly ConcurrentQueue<SocketSender> _queue = new();
+        private readonly PipeScheduler _scheduler;
+        private bool _disposed;
+
+        public SocketSenderPool(PipeScheduler scheduler)
+        {
+            _scheduler = scheduler;
+        }
+
+        public SocketSender Rent()
+        {
+            if (_queue.TryDequeue(out var sender))
+            {
+                return sender;
+            }
+            return new SocketSender(_scheduler);
+        }
+
+        public void Return(SocketSender sender)
+        {
+            if (_disposed || _queue.Count > MaxQueueSize)
```

Comment:
Isn't checking queue.Count > MaxQueueSize thread unsafe? From the way this code looks, we expect multiple calls to Rent and Return at the same time. This code may cause the queue to grow larger that 1024. This may be fine honestly.

Reply:
Yea but it doesn't matter. There's no hard limit

Reply:
Even though we don't need a hard limit, we should track the count independently for perf/predictability since we query it so frequently.

[ConcurrentQueue.Count](https://github.com/dotnet/runtime/blob/ec5d2eeade1ba4f203b2166c399056e124b4cfb8/src/libraries/System.Private.CoreLib/src/System/Collections/Concurrent/ConcurrentQueue.cs#L246-L349) is pretty harry. It takes a lock and loops if there are more than 2 segments. Based on my reading of the code, there will be more than 2 segments if the pool ever reaches more than 92 items.

Reply:
But does it really matter?

Reply:
Are you thinking it'll be a performance problem or a correctness problem?

@stephentoub we really need to expose ConcurrentQueueSegment\<T\> for pooling scenarios like this.

Reply:
My concern is purely perf.

Reply:
https://github.com/dotnet/runtime/issues/23700

Reply:
😢 . Do you think it's reasonable to revisit? The concurrent queue with a bounded capacity seemed so simple. I don't think holding off for an object pool (which comes with policy) should overlap.